explorer/interpreter/matching_impl_set.h

```diff
@@ -0,0 +1,121 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef CARBON_EXPLORER_INTERPRETER_MATCHING_IMPL_SET_H_
+#define CARBON_EXPLORER_INTERPRETER_MATCHING_IMPL_SET_H_
+
+#include <vector>
+
+#include "common/ostream.h"
+#include "explorer/ast/declaration.h"
+#include "explorer/common/nonnull.h"
+#include "explorer/interpreter/impl_scope.h"
+#include "explorer/interpreter/value.h"
+#include "llvm/ADT/DenseMap.h"
+
+namespace Carbon {
+
+// A set of impl matches that we're currently performing. Each `Match`
+// represents an attempt to match a `type as interface` query against an `impl`
+// declaration. This is used to detect and reject non-termination when impl
+// matching recursively triggers further impl matching.
+//
+// The approach we use is to count the number of times each "label" appears
+// within the type and interface, where a label is the name of a declared
+// entity such as a class or interface, or a primitive like `type` or `bool`.
+// For example, `Optional(i32*)` contains the labels for `Optional`, `i32`, and
+// `*` (built-in pointer type) once each.
+//
+// If we ever try matching the same `impl` twice, where the inner match
+// contains at least as many appearances of each label as the outer match, we
+// reject. We also reject if a query results in the exact same query being
+// performed again.
```

Comment:
```suggestion
// contains at least as many appearances of each label as the outer match, `Match::Check` will error. It will also error if a query results in the exact same query being
// performed again.
```

Trying to be clear what "reject" means.

Reply:
(As per the previous comment, I was trying to describe the language rule here, rather than the implementation strategy, so I don't want to talk about errors or particular function calls here. Hopefully the comment update makes this clearer.)